File: runtime/java/Json.java
public final class Json {

	private Json() {
	}

	public static String toJson(Object obj) {
		IdentityHashMap<Object, Boolean> seen = new IdentityHashMap<>();
		StringBuilder sb = new StringBuilder(256);
		write(obj, sb, seen);
		return sb.toString();
	}

	private static void write(Object obj, StringBuilder sb, IdentityHashMap<Object, Boolean> seen) {
		if (obj == null) {
			sb.append("null");
			return;
		}
		if (obj instanceof Boolean) {
			sb.append(((Boolean) obj) ? "true" : "false");
			return;
		}
		if (obj instanceof Number) {
			sb.append(obj);
			return;
		}
		if (obj instanceof Character) {
			quote(String.valueOf(obj), sb);
			return;
		}
		if (obj instanceof CharSequence) {
			quote(obj.toString(), sb);
			return;
		}

		if (obj instanceof Enum) {
			quote(((Enum<?>) obj).name(), sb);
			return;
		}

		if (obj instanceof TemporalAccessor) {
			quote(obj.toString(), sb);
			return;
		}
		if (obj instanceof Date) {
			quote(String.valueOf(((Date) obj).getTime()), sb);
			return;
		}
		if (seen.put(obj, Boolean.TRUE) != null) {
			throw new IllegalArgumentException("Cycle detected while encoding to json");
		}
		try {
			if (obj instanceof Map<?, ?>) {
				sb.append('{');
				boolean first = true;
				for (Map.Entry<?, ?> e : ((Map<?, ?>) obj).entrySet()) {
					if (!first) {
						sb.append(',');
					}
					first = false;
					String key = String.valueOf(e.getKey());
					quote(key, sb);
					sb.append(':');
					write(e.getValue(), sb, seen);
				}
				sb.append('}');
				return;
			}
			if (obj instanceof Iterable<?>) {
				sb.append('[');
				boolean first = true;
				for (Object x : (Iterable<?>) obj) {
					if (!first) {
						sb.append(',');
					}
					first = false;
					write(x, sb, seen);
				}
				sb.append(']');
				return;
			}
			Class<?> c = obj.getClass();
			if (c.isArray()) {
				sb.append('[');
				int n = Array.getLength(obj);
				for (int i = 0; i < n; i++) {
					if (i > 0) {
						sb.append(',');
					}
					write(Array.get(obj, i), sb, seen);
				}
				sb.append(']');
				return;
			}
			sb.append('{');
			boolean first = true;
			for (Field f : allFields(c)) {
				int m = f.getModifiers();
				if (Modifier.isStatic(m) || Modifier.isTransient(m) || f.isSynthetic()) {
					continue;
				}
				String name = f.getName();
				if (name.startsWith("_")) {
					continue;
				}
				Object v;
				try {
					f.setAccessible(true);
					v = f.get(obj);
				}
				catch (Exception ex) {
					continue;
				}

				if (!first) {
					sb.append(',');
				}
				first = false;
				quote(name, sb);
				sb.append(':');
				write(v, sb, seen);
			}
			sb.append('}');
		}
		finally {
			seen.remove(obj);
		}
	}

	private static List<Field> allFields(Class<?> c) {
		ArrayList<Field> out = new ArrayList<>();
		for (Class<?> x = c; x != null && x != Object.class; x = x.getSuperclass()) {
			Field[] fs = x.getDeclaredFields();
			Collections.addAll(out, fs);
		}
		return out;
	}

	private static void quote(String s, StringBuilder sb) {
		sb.append('"');
		for (int i = 0; i < s.length(); i++) {
			char ch = s.charAt(i);
			switch (ch) {
				case '"':
					sb.append("\\\"");
					break;
				case '\\':
					sb.append("\\\\");
					break;
				case '\b':
					sb.append("\\b");
					break;
				case '\f':
					sb.append("\\f");
					break;
				case '\n':
					sb.append("\\n");
					break;
				case '\r':
					sb.append("\\r");
					break;
				case '\t':
					sb.append("\\t");
					break;
				default:
					if (ch < 0x20) {
						sb.append(String.format("\\u%04x", (int) ch));
					}
					else {
						sb.append(ch);
					}
			}
		}
		sb.append('"');
	}
}

File: runtimes/kotlin/Json.kt
object Json {
    fun toJson(obj: Any?): String {
        val seen = IdentityHashMap<Any, Boolean>()
        val sb = StringBuilder(256)
        write(obj, sb, seen)
        return sb.toString()
    }

    private fun write(obj: Any?, sb: StringBuilder, seen: IdentityHashMap<Any, Boolean>) {
        if (obj == null) {
            sb.append("null")
            return
        }
        if (obj is Boolean) {
            sb.append(if (obj) "true" else "false")
            return
        }
        if (obj is Number) {
            sb.append(obj)
            return
        }
        if (obj is Char) {
            quote(obj.toString(), sb)
            return
        }
        if (obj is CharSequence) {
            quote(obj.toString(), sb)
            return
        }
        if (obj is Enum<*>) {
            quote(obj.name, sb)
            return
        }
        if (obj is TemporalAccessor) {
            quote(obj.toString(), sb)
            return
        }
        if (obj is Date) {
            quote(obj.time.toString(), sb)
            return
        }
        if (seen.put(obj, true) != null) throw IllegalArgumentException("Cycle detected")
        try {
            if (obj is Map<*, *>) {
                sb.append('{')
                var first = true
                for ((k, v) in obj) {
                    if (!first) sb.append(',')
                    first = false
                    quote(k.toString(), sb)
                    sb.append(':')
                    write(v, sb, seen)
                }
                sb.append('}')
                return
            }
            if (obj is Iterable<*>) {
                sb.append('[')
                var first = true
                for (x in obj) {
                    if (!first) sb.append(',')
                    first = false
                    write(x, sb, seen)
                }
                sb.append(']')
                return
            }
            if (obj.javaClass.isArray) {
                sb.append('[')
                val n = Array.getLength(obj)
                for (i in 0 until n) {
                    if (i > 0) sb.append(',')
                    write(Array.get(obj, i), sb, seen)
                }
                sb.append(']')
                return
            }
            sb.append('{')
            var first = true
            for (f in allFields(obj.javaClass)) {
                val m = f.modifiers
                if (Modifier.isStatic(m) || Modifier.isTransient(m) || f.isSynthetic) continue
                if (f.name.startsWith("_")) continue
                f.isAccessible = true
                val v =
                    try {
                        f.get(obj)
                    } catch (e: Exception) {
                        continue
                    }
                if (!first) sb.append(',')
                first = false
                quote(f.name, sb)
                sb.append(':')
                write(v, sb, seen)
            }
            sb.append('}')
        } finally {
            seen.remove(obj)
        }
    }

    private fun allFields(c: Class<*>): List<Field> {
        val out = mutableListOf<Field>()
        var x: Class<*>? = c
        while (x != null && x != Any::class.java) {
            out.addAll(x.declaredFields)
            x = x.superclass
        }
        return out
    }

    private fun quote(s: String, sb: StringBuilder) {
        sb.append('"')
        for (ch in s) {
            when (ch) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\b' -> sb.append("\\b")
                '\u000C' -> sb.append("\\f")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                else ->
                    if (ch < '\u0020') sb.append(String.format("\\u%04x", ch.code))
                    else sb.append(ch)
            }
        }
        sb.append('"')
    }
}
